From f0d5f156063ae8178da8c7439f8b547b8492d2be Mon Sep 17 00:00:00 2001
From: Matheus Sampaio Queiroga <srherobrine20@gmail.com>
Date: Fri, 17 Jul 2026 20:10:24 +0000
Subject: [PATCH 091/160] net: airoha: fix EN7523 CDM2 OAM queue selection

EN7523 uses bits 15:12 of CDM2_FWD_CFG for the PSE OAM queue
selection. The high queue-selection field used by later Airoha SoCs
does not route GPON OMCI traffic on EN7523.

Program the EN7523-specific field so downstream OMCI frames are sent
to QDMA RX ring 15.

Signed-off-by: Matheus Sampaio Queiroga <srherobrine20@gmail.com>
---
 drivers/net/ethernet/airoha/airoha_eth.c  | 3 +++
 drivers/net/ethernet/airoha/airoha_regs.h | 2 +-
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/airoha/airoha_eth.c b/drivers/net/ethernet/airoha/airoha_eth.c
index 110d574d9de8..299159d30b11 100644
--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -1027,6 +1027,9 @@ static int airoha_fe_init(struct airoha_eth *eth)
 	airoha_fe_set(eth, REG_GDM_MISC_CFG,
 		      GDM2_RDM_ACK_WAIT_PREF_MASK |
 		      GDM2_CHN_VLD_MODE_MASK);
+	/* EN7523 routes PSE OAM through bits 15:12. Later SoCs use
+	 * the high queue-selection field instead.
+	 */
 	airoha_fe_rmw(eth, REG_CDM_FWD_CFG(2),
 		      airoha_is(eth, en7523) ? EN7523_CDM_OAM_QSEL_MASK : CDM_OAM_QSEL_MASK,
 		      airoha_is(eth, en7523) ? FIELD_PREP(EN7523_CDM_OAM_QSEL_MASK, 15) :
diff --git a/drivers/net/ethernet/airoha/airoha_regs.h b/drivers/net/ethernet/airoha/airoha_regs.h
index d6af4db847bb..b0aacfa63cd3 100644
--- a/drivers/net/ethernet/airoha/airoha_regs.h
+++ b/drivers/net/ethernet/airoha/airoha_regs.h
@@ -131,7 +131,7 @@
 
 #define REG_CDM_FWD_CFG(_n)		(CDM_BASE(_n) + 0x08)
 #define CDM_OAM_QSEL_MASK		GENMASK(31, 27)
-#define EN7523_CDM_OAM_QSEL_MASK	GENMASK(31, 28)
+#define EN7523_CDM_OAM_QSEL_MASK	GENMASK(15, 12)
 #define CDM_VIP_QSEL_MASK		GENMASK(24, 20)
 #define EN7523_CDM_VIP_QSEL_MASK	GENMASK(27, 24)
 
-- 
2.55.0

